alacritty_terminal/src/graphics/sixel.rs

```diff
@@ -0,0 +1,772 @@
+
+        // At this point, we know that the new width is greater than the
+        // current one, so we have to extend the buffer and move the rows to
+        // their new positions.
+        let height = usize::max(height, self.height);
+
+        self.pixels.resize(height * width, self.background);
+
+        for y in (0..self.height).rev() {
+            for x in (0..self.width).rev() {
+                let old = y * self.width + x;
+                let new = y * width + x;
+                self.pixels.swap(old, new);
+            }
+        }
+
+        self.width = width;
+        self.height = height;
+        Ok(())
+    }
+
+    /// Add a sixel using the selected color register, and move the active
+    /// position.
+    fn add_sixel(&mut self, sixel: Sixel) -> Result<(), Error> {
+        self.finish_command()?;
+
+        // Take the repeat count and reset it.
+        //
+        // `max` function is used because the Sixel reference specifies
+        // that a repeat count of zero implies a repeat count of 1.
+        let repeat = max(1, mem::take(&mut self.repeat_count));
+
+        self.ensure_size(self.x + repeat, self.y + sixel.height())?;
+
+        if sixel.0 != 0 {
+            let mut index = self.width * self.y + self.x;
+            for dot in sixel.dots() {
+                if dot {
+                    for pixel in &mut self.pixels[index..index + repeat] {
+                        *pixel = self.selected_color_register;
+                    }
+                }
+
+                index += self.width;
+            }
+        }
+
+        self.x += repeat;
+
+        Ok(())
+    }
+
+    /// Returns the final graphic to append to the grid, with the palette
+    /// built in the process.
+    pub fn finish(mut self) -> Result<(GraphicData, Vec<Rgb>), Error> {
+        self.finish_command()?;
+
+        trace!(
+            "Finish Sixel parser: width={}, height={}, color_registers={}",
+            self.width,
+            self.height,
+            self.color_registers.len()
+        );
+
+        let mut rgba_pixels = Vec::with_capacity(self.pixels.len() * 4);
+
+        for &register in &self.pixels {
+            let pixel = {
+                if register == REG_TRANSPARENT {
+                    [0; 4]
+                } else {
+                    match self.color_registers.get(register.0 as usize) {
+                        None => [0, 0, 0, 255],
+                        Some(color) => [color.r, color.g, color.b, 255],
+                    }
+                }
+            };
+
+            rgba_pixels.extend_from_slice(&pixel);
+        }
+
+        let data = GraphicData {
+            id: GraphicId(0),
+            height: self.height,
+            width: self.width,
+            color_type: ColorType::Rgba,
+            pixels: rgba_pixels,
+        };
+
+        Ok((data, self.color_registers))
+    }
+}
+
+/// Compute a RGB value from HLS.
+///
+/// Input and output values are in the range of `0..=100`.
```

Comment:
According to https://archive.org/details/bitsavers_decstandar0VideoSystemsReferenceManualDec91_74264381/page/n927/mode/2up, hue is in the range of `0..360`